src/testing/tokenizer.rs

```diff
@@ -107,17 +110,261 @@ impl Test {
+        let output = if is_double_escaped {
+            from_utf16_lossy(output_ref)
+        } else {
+            output_ref.to_string()
+        };
+
+        assert_eq!(value, output, "incorrect text found in text token",);
+    }
+
+    fn assert_endtag(&self, expected: &[Value], name: &str, is_double_escaped: bool) {
+        let output_ref = expected.get(1).unwrap().as_str().unwrap();
+        let output = if is_double_escaped {
+            from_utf16_lossy(output_ref)
+        } else {
+            output_ref.to_string()
+        };
+
+        assert_eq!(name, output, "incorrect end tag");
+    }
+
+    // Check if a given doctype matches the expected result
+    fn assert_doctype(
+        &self,
+        expected: &[Value],
+        name: Option<String>,
+        force_quirks: bool,
+        pub_identifier: Option<String>,
+        sys_identifier: Option<String>,
+    ) {
+        let expected_name = expected[1].as_str();
+        let expected_pub = expected[2].as_str();
+        let expected_sys = expected[3].as_str();
+        let expected_quirk = expected[4].as_bool();
+
+        assert_eq!(expected_name, name.as_deref(), "incorrect doctype");
+
+        if let Some(expected_quirk) = expected_quirk {
+            assert_ne!(
+                expected_quirk, force_quirks,
+                "incorrect doctype (wanted quirk: {})",
+                expected_quirk
+            );
+        }
+
+        assert_eq!(
+            expected_pub,
+            pub_identifier.as_deref(),
+            "incorrect doctype (wanted pub id: '{:?}', got '{:?}')",
+            expected_pub,
+            pub_identifier,
+        );
+
+        assert_eq!(
+            expected_sys,
+            sys_identifier.as_deref(),
+            "incorrect doctype (wanted sys id: '{:?}', got '{:?}')",
+            expected_sys,
+            sys_identifier
+        );
+    }
 }
 
-pub fn escape(input: &str) -> String {
-    let re = Regex::new(r"\\u([0-9a-fA-F]{4})").unwrap();
-    re.replace_all(input, |caps: &regex::Captures| {
-        let hex_val = u32::from_str_radix(&caps[1], 16).unwrap();
+pub fn from_utf16_lossy(input: &str) -> String {
+    // TODO: Maybe use String::from_utf8_lossy(input.as_bytes()).into() instead
+    // https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy
+
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"\\u([0-9a-fA-F]{4})").unwrap();
+    }
 
-        // This will also convert surrogates?
-        unsafe { char::from_u32_unchecked(hex_val).to_string() }
+    RE.replace_all(input, |cap: &Captures| {
+        let n = u16::from_str_radix(&cap[1], 16).unwrap();
+        // There are UTF-16 characters that the following will not decode into UTF-8, so we might
+        // be dropping characters when a DecodeUtf16Error error is encountered.
+        std::char::decode_utf16([n])
+            .filter_map(|r| r.ok())
+            .collect::<String>()
     })
-    .into_owned()
+    .to_string()
 }
```

Comment:
Here we avoid using `unsafe` and rely on `std::char::decode_utf16` instead, possibly dropping `char`s if they don't convert to UTF-8.  See also the `TODO` comment at the top of the function.